支持Release/v2.6 运行在Zephyr RTOS上

cocoeoli opened the pull request:
我们将paddle lite作为zephyr RTOS的一个第三方外部模块，加入了zephyr RTOS的代码仓库中。并在Firefly ROC-RK3568-PC开发板上进行了测试，预测结果和arm linux一致，同时在部分场景下paddle lite on zephyr RTOS的推理速度优于linux.

hong19860320 commented on lite/api/light_api.cc:
```diff
@@ -43,7 +45,7 @@ void LightPredictor::Build(const std::string& model_dir,
                            bool model_from_memory) {
   switch (model_type) {
-#ifndef LITE_ON_TINY_PUBLISH
+#if !defined(LITE_ON_TINY_PUBLISH) && !defined(CONFIG_LITE_WITH_ZEPHYR)
```
CONFIG_LITE_WITH_ZEPHYR
改成
LITE_WITH_ZEPHYR

下同

cocoeoli replied:
这个`CONFIG_`头是Zephyr操作系统中必须的，不影响Paddlite端代码。

cocoeoli replied:
@hong19860320 这种可以接受吗？

hong19860320 replied:
不太能接受，需要满足 LITE 中宏定义的规范，我理解可以在 cmake 里增加一下 LITE_WITH_ZEPHYR 宏。

cocoeoli replied:
已经在zephyr操作系统端进行修改，增加了 LITE_WITH_ZEPHYR 宏

hong19860320 commented on lite/backends/arm/math/saturate.h:
```diff
@@ -17,6 +17,9 @@
+#endif
 
 namespace paddle {
 namespace lite {
```
100755 → 100644 这些文件修改不要提交

cocoeoli replied:
100755 → 100644：这些文件都已经从commit删除

hong19860320 commented on lite/core/context.h:
```diff
@@ -247,6 +247,51 @@ class Context<TargetType::kARM> {
 };
 #endif
 
+#ifdef CONFIG_LITE_WITH_ZEPHYR
```
代码冗余，为什么不能复用 class Context<TargetType::kARM> ？

cocoeoli replied:
已经删除对应的冗余代码

hong19860320 commented on lite/model_parser/model_parser.cc:
```diff
@@ -835,7 +890,56 @@ void LoadModelNaiveFromFile(const std::string &filename,
 }
+#endif
+
+#ifdef CONFIG_LITE_WITH_ZEPHYR
```
为什么不能复用之前的实现？建议尽量减少代码冗余。

cocoeoli replied:
已经删除对应的冗余代码

hong19860320 commented on lite/model_parser/naive_buffer/naive_buffer.cc:
```diff
@@ -54,6 +54,39 @@ void BinaryTable::AppendToFile(const std::string &filename) const {
 }
 
+void BinaryTable::LoadFromFile(  uint64_t  &topo_size,uint8_t* start)
+{
```
代码风格需要注意，请按照文档的要求安装 clang-format hook https://www.paddlepaddle.org.cn/documentation/docs/zh/dev_guides/code_contributing_path_cn.html

cocoeoli replied:
代码风格检查已经按照clang-format hook修改：
![image](https://github.com/PaddlePaddle/Paddle-Lite/assets/40254934/09989008-bb93-4293-b6e8-c3bb27cc7216)

hong19860320 commented on lite/backends/arm/math/dotprod/gemm_sdot.h:
```diff
@@ -0,0 +1,369 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
为什么需要增加这个代码？

cocoeoli replied:
已删除了该代码，代码放错了位置

hong19860320 commented on lite/model_parser/model_parser.cc:
```diff
@@ -693,6 +711,42 @@ void LoadCombinedParamsNaive(const std::string &path,
   }
 }
 
+void LoadCombinedParamsNaive(const std::string &path,
```
为什么需要单独定义一个这样的函数，是Zephyr 上有什么限制吗？我看和上面 LoadCombinedParamsNaive 的实现除了细微的差异，其它基本上一致，是否能降低代码冗余？

cocoeoli replied:
由于Zephyr文件系统不够完善，像fopen这类的函数无法使用。所以直接通过内存的方式来读取模型文件才出现下列代码。首先将模型映射到指定地址上面，让然后进行字节复制来读取模型文件的内容。paddlelite模型包含五个部分，第一第二部分是版本号信息，第三部分包含拓扑结构包含多少个字节，而第四部分就是拓扑结构，第五部分是参数字段。而LoadCombinedParamsNaive就是读取第五部分，但是第五部分的字节数需要根据前面几个部分读了多少个字节，然后根据首地址来读第五部分。

hong19860320 replied:
嗯，了解了，是不是可以在Zephyr+PaddleLite 强制使用从内存读取模型的接口？
set_model_buffer：https://github.com/PaddlePaddle/Paddle-Lite/blob/545c04ee9429405b496bcd73768fb5e1595e7a56/lite/api/paddle_api.h#L613C8-L613C29 
api 文档：https://www.paddlepaddle.org.cn/lite/v2.12/api_reference/cxx_api_doc.html#set-model-from-buffer
这样可减少对 lite/model_parser/model_parser.cc 做太定制化的修改。

cocoeoli replied:
zephyr下相对路径绝对路径这种东西并不存在，无法通过文件路径方式找到模型。通过内存读取模型的方式指的是需要知道模型所在的内存虚拟地址，并且zephyr下完成物理地址到虚拟地址1比1映射。所以强制使用从内存读取模型的接口不行。针对对model parser 和 naive_buffer 基础模块的定制化修改，可以将定制化的代码放入到zephyr当中，减少了对原paddlelite的修改。

hong19860320 commented on lite/model_parser/model_parser.cc:
```diff
@@ -835,7 +890,56 @@ void LoadModelNaiveFromFile(const std::string &filename,
+  cpp_prog->ClearBlocks();
+  // ModelFile
+  const std::string prog_path = filename;
+  uint8_t* start = (uint8_t*)0x70000000;
```
为什么有这么多 hardcode？ 是否能减少这类入侵式的修改？

cocoeoli replied:
已去除hardcode

hong19860320 commented on lite/model_parser/naive_buffer/naive_buffer.cc:
```diff
@@ -54,6 +54,39 @@ void BinaryTable::AppendToFile(const std::string &filename) const {
+  uint64_t total = 0;
+  for(i=0;i<topo_size;i++)
+  {
+    memcpy(reinterpret_cast<char *>(address), start, 1);
```
这个修改是不是只适合Zephyr？

cocoeoli replied:
对，因为zephyr使用内存模型读取的方式

hong19860320 commented on lite/model_parser/naive_buffer/naive_buffer.cc:
```diff
@@ -54,6 +54,34 @@ void BinaryTable::AppendToFile(const std::string &filename) const {
+}
+#endif
+
 void BinaryTable::LoadFromFile(const std::string &filename,
```
同上，不建议对 model parser 和 naive_buffer 基础模块进行太定制化的修改。

cocoeoli replied:
针对对model parser 和 naive_buffer 基础模块的定制化修改，将定制化的代码放入到zephyr当中，减少了对原paddlelite的修改。

cocoeoli committed:
RTOS support: support paddle lite run on zephyr rtos

We integrated Paddle Lite as a third-party external module into the Zephyr RTOS code repository.
Subsequently, we conducted tests on the Firefly ROC-RK3568-PC development board.
The predictive outcomes aligned with those of ARM Linux. Additionally, in some scenarios,
the speed of Paddle Lite on Zephyr RTOS surpassed that of Linux.

Signed-off-by: Charlie Xiong <1981639884@qq.com>

hong19860320 reviewed (approved):
LGTM，已合入。

ZhijianZhou01 commented:
The RTOS-enabled paddle lite looks like an interesting job.

onecatcn commented:
Congratulations to the hnu-esnl team for being the first to merge RTOS into the upstream of Paddle Lite! This achievement enables Paddle Lite to run on the open-source RTOS, Zephyr. Thank you for the contribution to Paddle-Lite community!

hlleng commented:
Zephyr is one of the most famous RTOS. This is a good job for Zephyr and Paddle Lite.
